File: ios/MiniBozor/Data/Repositories/CatalogRepository.swift
import Foundation

/// Screens 07–16 and 35.
struct CatalogRepository {
    private let api = APIClient.shared

    func home(city: String) async -> Outcome<HomeDTO> {
        await run { try await api.get("home", query: [URLQueryItem(name: "city", value: city)]) }
    }

    func rootCategories() async -> Outcome<[CategoryDTO]> {
        await run { try await api.get("categories") }
    }

    func children(of slug: String) async -> Outcome<[CategoryDTO]> {
        await run { try await api.get("categories", query: [URLQueryItem(name: "parent", value: slug)]) }
    }

    func category(_ slug: String) async -> Outcome<CategoryDTO> {
        await run { try await api.get("categories/\(slug)") }
    }

    func products(_ query: ProductQuery, page: Int) async -> Outcome<PageDTO<ProductCardDTO>> {
        await run { try await api.get("products", query: query.queryItems(page: page)) }
    }

    func filters(category: String?) async -> Outcome<FiltersDTO> {
        let items = category.map { [URLQueryItem(name: "category", value: $0)] } ?? []
        return await run { try await api.get("products/filters", query: items) }
    }

    func product(_ id: Int) async -> Outcome<ProductDTO> {
        await run { try await api.get("products/\(id)") }
    }

    func similar(to id: Int) async -> Outcome<[ProductCardDTO]> {
        await run { try await api.get("products/\(id)/similar") }
    }

    func searchLanding() async -> Outcome<SearchLandingDTO> {
        await run { try await api.get("search") }
    }

    func suggest(_ text: String) async -> Outcome<[SuggestionDTO]> {
        await run { try await api.get("search/suggest", query: [URLQueryItem(name: "q", value: text)]) }
    }

    func rememberSearch(_ text: String) async {
        let _: MessageDTO? = try? await api.post(
            "search/recent",
            query: [URLQueryItem(name: "query", value: text)]
        )
    }

    func clearSearchHistory() async {
        let _: MessageDTO? = try? await api.delete("search/recent")
    }

    func reviewSummary(productId: Int) async -> Outcome<ReviewSummaryDTO> {
        await run { try await api.get("products/\(productId)/reviews/summary") }
    }

    func reviews(productId: Int, stars: Int?, page: Int) async -> Outcome<PageDTO<ReviewDTO>> {
        var items = [URLQueryItem(name: "page", value: String(page))]
        if let stars { items.append(URLQueryItem(name: "stars", value: String(stars))) }
        return await run { try await api.get("products/\(productId)/reviews", query: items) }
    }

    func reviewTags() async -> Outcome<[String]> {
        await run { try await api.get("reviews/tags") }
    }

    func createReview(productId: Int, body: ReviewCreateRequest) async -> Outcome<ReviewDTO> {
        await run { try await api.post("products/\(productId)/reviews", body: body) }
    }

    func likeReview(_ id: Int) async -> Outcome<ReviewDTO> {
        await run { try await api.post("reviews/\(id)/like") }
    }

    func favorites(page: Int = 1) async -> Outcome<PageDTO<ProductCardDTO>> {
        await run { try await api.get("favorites", query: [URLQueryItem(name: "page", value: String(page))]) }
    }

    func setFavorite(productId: Int, favorite: Bool) async -> Outcome<Void> {
        await run {
            if favorite {
                let _: MessageDTO = try await api.put("favorites/\(productId)")
            } else {
                let _: MessageDTO = try await api.delete("favorites/\(productId)")
            }
        }
    }
}

File: android/app/src/main/java/uz/minibozor/data/repository/CatalogRepository.kt
package uz.minibozor.data.repository

import uz.minibozor.core.util.Outcome
import uz.minibozor.core.util.apiCall
import uz.minibozor.data.remote.MiniBozorApi
import uz.minibozor.data.remote.dto.*
import uz.minibozor.ui.catalog.ProductQuery
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CatalogRepository @Inject constructor(private val api: MiniBozorApi) {

    suspend fun home(city: String): Outcome<HomeDto> = apiCall { api.home(city) }

    suspend fun rootCategories(): Outcome<List<CategoryDto>> = apiCall { api.categories() }

    suspend fun children(parent: String): Outcome<List<CategoryDto>> =
        apiCall { api.categories(parent = parent) }

    suspend fun category(slug: String): Outcome<CategoryDto> = apiCall { api.category(slug) }

    suspend fun products(query: ProductQuery, page: Int): Outcome<PageDto<ProductCardDto>> =
        apiCall {
            api.products(
                q = query.text,
                category = query.category,
                brand = query.brands.toList(),
                minPrice = query.minPrice,
                maxPrice = query.maxPrice,
                minRating = query.minRating,
                size = query.sizes.toList(),
                nextDayDelivery = query.flags["next_day_delivery"],
                freeDelivery = query.flags["free_delivery"],
                discounted = query.flags["discounted"],
                isOriginal = query.flags["is_original"],
                // Only ever sent when it is on: the server leaves them out by
                // default, and a `false` on every request is noise.
                showSoldOut = true.takeIf { query.showSoldOut },
                sort = query.sort,
                page = page,
            )
        }

    suspend fun filters(category: String?): Outcome<FiltersDto> = apiCall { api.filters(category) }

    suspend fun product(id: Int): Outcome<ProductDto> = apiCall { api.product(id) }

    suspend fun similar(id: Int): Outcome<List<ProductCardDto>> = apiCall { api.similar(id) }

    suspend fun searchLanding(): Outcome<SearchLandingDto> = apiCall { api.searchLanding() }

    suspend fun suggest(q: String): Outcome<List<SuggestionDto>> = apiCall { api.suggest(q) }

    suspend fun rememberSearch(q: String) = apiCall { api.rememberSearch(q) }

    suspend fun clearSearchHistory() = apiCall { api.clearSearchHistory() }

    suspend fun reviewSummary(productId: Int): Outcome<ReviewSummaryDto> =
        apiCall { api.reviewSummary(productId) }

    suspend fun reviews(productId: Int, stars: Int?, page: Int): Outcome<PageDto<ReviewDto>> =
        apiCall { api.reviews(productId, stars, page = page) }

    suspend fun reviewTags(): Outcome<List<String>> = apiCall { api.reviewTags() }

    suspend fun createReview(productId: Int, body: ReviewCreateRequest): Outcome<ReviewDto> =
        apiCall { api.createReview(productId, body) }

    suspend fun likeReview(reviewId: Int): Outcome<ReviewDto> = apiCall { api.likeReview(reviewId) }

    suspend fun favorites(page: Int): Outcome<PageDto<ProductCardDto>> =
        apiCall { api.favorites(page) }

    suspend fun setFavorite(productId: Int, favorite: Boolean): Outcome<Unit> =
        apiCall { if (favorite) api.addFavorite(productId) else api.removeFavorite(productId) }
            .let { if (it is Outcome.Failure) it else Outcome.Success(Unit) }
}
